### DISCUSSION TITLE
Progress bar fixed to bottom of window

### DISCUSSION THREAD

**Discussion by garlontas** [Textualize/rich#1571] [Q&A] [answered]
Hello,
is there a possibility to create a progress bar fixed to the bottom of the terminal window, like in this image?

![graphic](https://user-images.githubusercontent.com/70283087/136953250-874879e1-6de8-4368-ab76-1478fa3d56a5.png)

I want the text to appear above the progress bar.

Thank you in advance!

**Comment by mrvkino** [comment, marked as answer]
Hi,

By default the progress bar in rich will be shown under the text being printed.

Here is a very simple code sample that will do what you want:
```python3
import time
from rich.progress import track
sample_text = [
    "Text may be printed while the progress bars are rendering.",
    "This is pretty neat.",
    "I do not have a lot of things to say.",
    "So I will stop here."
]
for example in track(sample_text, total=len(sample_text)):
    time.sleep(1)
    print(example)
```

[rich](https://github.com/willmcgugan/rich#installing) also has a more advance example you can run:
```python
python -m rich.progress
```

This is a more feature rich solution that you can look at [here](https://github.com/willmcgugan/rich/blob/256697915d9676b7d5cb6abd5559132ff5f92fdf/rich/progress.py#L967) (😜pun intended).

Here is the quick rundown:

- **line 977 to 1006**: The data being printed is generated from . 
- **line 1010**: it makes the list an infinite iterable.
- **line 1014**: it setup the context to run the progress bar with various columns (information about the process being tracked).
- **line 1029 to 1034**: The `while` loop is the main loop updating the progress of each progress bar and printing the data above at random interval. It is also printing to the `stdout` using the `log` method.

I am pretty sure you can skip line 1012 and not give a value to the `console` argument at line 1021. The default is to print to `stdout`.

Hope this helps.

**Reply by willmcgugan** [reply]
Thanks @mrvkino 

@garlontas This is the default behaviour. Enjoy!

**Reply by garlontas** [reply]
Sorry, I think I didn't say clearly what I need :disappointed: .
Your answer is similar to what I want, there is just one problem:
Except text, I want to display also text prompts with this library: https://github.com/kazhala/InquirerPy
I adapted your code and wrote this:
```python
import time
from pathlib import Path

from InquirerPy import inquirer
from InquirerPy.validator import PathValidator
from rich.progress import track

sample_text = [
    "Text may be printed while the progress bars are rendering.",
    "This is pretty neat.",
    "I do not have a lot of things to say.",
    "So I will stop here."
]
for example in track(sample_text, total=len(sample_text)):
    time.sleep(1)
    print(example)
    inquirer.filepath(
        message="Enter file to upload:",
        default=str(Path.cwd()),
        validate=PathValidator(is_file=True, message="Input is not a file"),
        only_files=True,
    ).execute()


```
Unfortunately I get this output:

https://user-images.githubusercontent.com/70283087/137155297-f38fd280-2e98-4cc9-99ad-a0ba34b460ff.mp4

So my text prompt can't get rendered because stdout is not a terminal :cry: 

Sorry again, and thanks for your previous answer!!

PS: @mrvkino I'm sorry you wrote so much, and I can't use it because I didn't write a detailed question. Thank you very much for trying to help!!!

**Reply by saschalalala** [reply]
@garlontas Regarding your question with the fixed progress, have a look at layouts or tables. You can position several rows/columns or panels in a flexible way and put the progress bar in the footer, for example like this:

```python
#!/usr/bin/env python3
import time
import datetime as dt

from rich.layout import Layout
from rich.live import Live
from rich.progress import Progress
from rich.panel import Panel

sample_text = [
    "Text may be printed while the progress bars are rendering.",
    "This is pretty neat.",
    "I do not have a lot of things to say.",
    "So I will stop here."
]

total = len(sample_text)
progress = Progress()
task = progress.add_task("task", total=total)

main_text = ""
now = str(dt.datetime.now())

def generate_layout():
    layout = Layout()
    layout.split(
        Layout(name="header", size=3),
        Layout(ratio=1, name="main"),
        Layout(ratio=1, name="datetime"),
        Layout(size=5, name="footer")
    )
    layout["datetime"].update(Panel(
        now,
        title="datetime",
        border_style="blue",
        padding=(2, 2),
    ))

    layout["main"].update(Panel(
        main_text,
        title="main",
        border_style="red",
        padding=(2, 2),
    ))

    layout["footer"].update(Panel(
        progress,
        title="Progress",
        border_style="green",
        padding=(2, 2),
    ))
    return layout


with Live(generate_layout(), screen=True) as live:
    for progress_counter, example in enumerate(sample_text):
        time.sleep(1)
        live.update(generate_layout())
        main_text = example * 10
        now = str(dt.datetime.now())
        progress.update(task, advance=1)
```

Regarding your second question about InquirerPy, I fiddled around a little and tried to render the inquirer .execute() output into a panel but I could not get it to work. Last thing I tried is something like:

```python
file_prompt = inquirer.filepath(
    message="Enter file to upload:",
    default=str('/tmp/foosadasd'),
    validate=PathValidator(is_file=True, message="Input is not a file"),
    only_files=False,
)
inquirer_console = Console()
inquirer_text = inquirer_console.render_str(file_prompt.execute())
layout["inquirer"].update(Panel(
    inquirer_text,
    title="main",
    border_style="red",
    padding=(2, 2),
))
```

but as soon as I do anything related to `file_prompt.execute()`, the loop is paused because the input is locked (same thing for the Prompt input of rich). I assume, something like this inside a live layout is just not possible.

**Reply by garlontas** [reply]
Thanks for your help @saschalalala :smile:  !!!
I will put the progress bar above the widgets … (I think it's the only solution)
Maybe I'll try to remove the InquirerPy text prompts after they are used by the user from the terminal.
I have to look how to do that.

Any ideas?
